### PR TITLE
Dwarf sourceline info

### PR DESCRIPTION
**Pull request by Sidhub723**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Added a handler in `ds_print_dwarf()` function in `librz/core/disasm.c`, if the variable asm.dwarf.lines is set to true. asm.dwarf.lines is set to false by default.

...

**Test plan**

...

**Closing issues**
This PR is with regards to [issue 653](https://github.com/rizinorg/rizin/issues/653)

...

### CODE REVIEW

**Comment by ret2libc** [comment]
Can you show here on GH how the output changes before and after your PR? What is the difference between `asm.dwarf` and `asm.dwarf.lines`?

**Comment by wargio** [comment]
add a test, please

**Comment by Sidhub723** [comment]
When only asm.dwarf is set to true and asm.dwarf.lines is set to false, below would be the disassembly output:
image
When asm.dwarf is still true, and line debug info display is enabled by setting asm.dwarf.lines to true, the disassembly changes to:
image
One thing i still remain unsure about it how to make the `Line Number -xx` be displayed in a color which matches the console theme.

File: librz/core/disasm.c [hunk 1]
```diff
@@ -349,7 +350,8 @@ static void ds_print_sysregs(RzDisasmState *ds);
 static void ds_print_fcn_name(RzDisasmState *ds);
 static void ds_print_as_string(RzDisasmState *ds);
 static bool ds_print_core_vmode(RzDisasmState *ds, int pos);
-static void ds_print_dwarf(RzDisasmState *ds);
+// static void ds_print_dwarf(RzDisasmState *ds);
```
Comment (XVilka) [inline librz/core/disasm.c]: Remove

File: librz/core/disasm.c [hunk 2]
```diff
@@ -3811,7 +3814,50 @@ static void ds_align_comment(RzDisasmState *ds) {
...
 
-static void ds_print_dwarf(RzDisasmState *ds) {
+static void ds_print_dwarf(RzCore *core, RzCmdStateOutput *state, RzDisasmState *ds) {
+
+	bool binFileExists = true;
+	bool SourceLineInfoExists = true;
+
+	RzBinFile *binfile = core->bin->cur;
+	if (!binfile || !binfile->o) {
+		// rz_cons_printf("No file loaded.\n");
```
Comment (XVilka) [inline librz/core/disasm.c]: Remove

File: librz/core/disasm.c [hunk 3]
```diff
@@ -3811,7 +3814,50 @@ static void ds_align_comment(RzDisasmState *ds) {
...
+		binFileExists = false;
+		// return false;
+	}
+	RzBinSourceLineInfo *li = binfile->o->lines;
+	if (!li) {
+		// rz_cons_printf("No Source line information available");
```
Comment (XVilka) [inline librz/core/disasm.c]: Same

File: librz/core/disasm.c [hunk 4]
```diff
@@ -3811,7 +3814,50 @@ static void ds_align_comment(RzDisasmState *ds) {
...
+		SourceLineInfoExists = false;
+		// return true;
+	}
+	if (ds->dwarfShowLines && ds->show_dwarf && SourceLineInfoExists && binFileExists) {
+
```
Comment (XVilka) [inline librz/core/disasm.c:3822]: Unnecessary empty line

File: librz/core/disasm.c [hunk 5]
```diff
@@ -3811,7 +3814,50 @@ static void ds_align_comment(RzDisasmState *ds) {
...
+		rz_cmd_state_output_array_start(state);
+		rz_cons_break_push(NULL, NULL);
+		RzBinSourceLineSample *temps = NULL;
+		for (size_t i = 0; i < li->samples_count; i++) {
+			if (rz_cons_is_breaked()) {
+				break;
+			}
+			temps = &li->samples[i];
+
```
Comment (XVilka) [inline librz/core/disasm.c]: Same

File: librz/core/disasm.c (hunk 4 above)
Comment (XVilka) [reply librz/core/disasm.c:3822]: You can extract that part into a separate static function for clarity

**Review by XVilka — CHANGES_REQUESTED** [review]
PLease add a test

**Comment by XVilka** [comment]
The overall format should be `filename:line`, e.g. `bla.c:315`, without `Line number...`, etc

**Comment by Sidhub723** [comment]
Will work on making these changes ASAP, thanks for the feedback!

**Comment by wargio** [comment]
also it should be like a comment so ` ; bla.c:315`

**Review by XVilka — CHANGES_REQUESTED** [review]
Please add a test, for both disabled and enabled states.

**Comment by XVilka** [comment]
@imbillow feel free to take over as part of your GSoC

**Comment by b1llow** [comment]
> @imbillow feel free to take over as part of your GSoC


What do you mean by 'take over'? Should I review the PR, or close it, open a new one and continue working from there?

**Comment by Sidhub723** [comment]
Sorry that I couldnt finish it quick. Goodluck to @imbillow !

**Comment by wargio** [comment]
> > @imbillow feel free to take over as part of your GSoC
> 
> What do you mean by 'take over'? Should I review the PR, or close it, open a new one and continue working from there?

take over

**Comment by b1llow** [comment]
Follow-up work in the new PR #3511

**Pull request closed without merging**